specification/dartLangSpec.tex

```diff
@@ -5452,42 +5452,93 @@ \subsection{Mixin Application}
 
 \LMHash{}%
-Let $S$ be a class,
-$M$ be a mixin with \NoIndex{required superinterface}s $T_1$, \ldots, $T_n$,
-\NoIndex{combined superinterface} $M_S$,
-\NoIndex{implemented interfaces} $I_1$, \ldots, $I_k$ and
-\metavar{members} as \NoIndex{mixin member declarations},
-and let $N$ be a name.
+\BlindDefineSymbol{S, S'}%
+Let $S$ be a class, and let $S'$ be a parameterized type of the form
+\code{$S$<$\cdots$>}.
+\commentary{%
+This includes the case where $S$ is non-generic and $S'$ is $S$.%
+}
+
+\LMHash{}%
+\BlindDefineSymbol{M, \metavar{members}}%
+Let $M$ be a mixin with member declarations \metavar{members}.
+\BlindDefineSymbol{M', T_j}%
+Let $M'$ be a parameterized type of the form
+\code{$M$<$\cdots$>}
+\commentary{(again including the case where $M'$ is $M$)}.
+Let \List{T}{1}{n} be the required superinterfaces,
+\BlindDefineSymbol{M_S, I_j}
+$M_S$ the combined superinterface,
+\List{I}{1}{k} the implemented interfaces,
+all of $M$ and corresponding to $M'$.
 
 \LMHash{}%
-It is a compile-time error to apply $M$ to $S$ if $S$ does not implement,
-directly or indirectly, all of $T_1$, \ldots, $T_n$.
+It is a compile-time error to apply $M'$ to $S'$
+unless $S'$ implements each of \List{T}{1}{n}
+(\ref{interfaceSuperinterfaces}).
 It is a compile-time error if any of \metavar{members} contains a
-super-invocation of a member $m$ \commentary{(for example \code{super.foo},
-\code{super + 2}, or \code{super[1] = 2})}, and $S$ does not have a concrete
-implementation of $m$ which is a valid override of the member $m$ in
-the interface $M_S$. \rationale{We treat super-invocations in mixins as
-interface invocations on the combined superinterface, so we require the
-superclass of a mixin application to have valid implementations of those
-interface members that are actually super-invoked.}
-
-\LMHash{}%
-The mixin application of $M$ to $S$ with name $N$ introduces a new
-class, $C$, with name $N$, superclass $S$,
-implemented interface $M$
-and \metavar{members} as instance members.
+super-invocation of a member $m$
+\commentary{%
+(for example \code{super.foo}, \code{super + 2}, or \code{super[1] = 2})%
+},
+and $S'$ does not have a concrete implementation of $m$ which is
+a valid override of the member $m$ in the interface $M_S$.
+\rationale{%
+We treat super-invocations in mixins as interface invocations on
+the combined superinterface,
+so we require the superclass of a mixin application to have
+valid implementations of those interface members
+that are actually super-invoked.%
+}
+
+\LMHash{}%
+Let \DefineSymbol{L_C} be the library containing the mixin application.
+\commentary{%
+That is, the library containing the clause \code{$S$ \WITH{} $M$}
+or the clause \code{$S_0$ \WITH{} $M_1$, \ldots,\ $M_k$, $M$} giving rise
+to the mixin application.%
+}
+Let \DefineSymbol{L_M} be the library containing the declaration of $M$.
+
+\LMHash{}%
+Assume that $S$ has a concrete member $m_S$ which is accessible to $L_M$,
```

Comment:
Did we agree on it needing to be *concrete*?

Reply:
No, this PR awaits the discussion in the language team: Do we want to support these mixin applications in all the cases where it does not give rise to an override, or do we want to raise an error on all clashes (abstract or concrete, that doesn't matter)?

My preferred approach is the former, because I don't want to eliminate expressive power gratuitously.

However, we certainly need to raise an error when there is an override because of an implicitly induced implementation (say, a stub that checks the type of an actual argument and calls `super.sameMember`).

This means that the text in this PR needs to be updated, but it should be updated in two different ways if we choose the former respectively the latter approach.

Reply:
Cf. the decision at the language meeting yesterday, the PR has now been updated to make it an error when the private collision as created by a mixin application, no matter what kind of clash we have (concrete/concrete, concrete/abstract, concrete/stub, abstract/abstract, it's always an error).